akka-stream/src/main/scala/akka/stream/impl/Buffers.scala

```diff
@@ -0,0 +1,204 @@
+      writeIdx += 1
+    }
+
+    protected def toOffset(idx: Int): Int
+
+    def put(idx: Int, elem: T): Unit = buffer(toOffset(idx)) = elem.asInstanceOf[AnyRef]
+    def get(idx: Int): T = buffer(toOffset(idx)).asInstanceOf[T]
+
+    def peek(): T = get(readIdx)
+
+    def dequeue(): T = {
+      val result = get(readIdx)
+      dropHead()
+      result
+    }
+
+    def clear(): Unit = {
+      java.util.Arrays.fill(buffer, null)
+      readIdx = 0
+      writeIdx = 0
+    }
+
+    def dropHead(): Unit = {
+      put(readIdx, null.asInstanceOf[T])
+      readIdx += 1
+    }
+
+    def dropTail(): Unit = {
+      writeIdx -= 1
+      put(writeIdx, null.asInstanceOf[T])
+    }
+  }
+
+  private final class ModuloFixedSizeBuffer[T](_size: Int) extends FixedSizeBuffer[T](_size) {
+    override protected def toOffset(idx: Int): Int = idx % capacity
+  }
+
+  private final class PowerOfTwoFixedSizeBuffer[T](_size: Int) extends FixedSizeBuffer[T](_size) {
+    private val Mask = capacity - 1
+    override protected def toOffset(idx: Int): Int = idx & Mask
+  }
+
+}
+
+/**
+ * INTERNAL API
+ */
+private[akka] final class BoundedBuffer[T](val capacity: Int) extends Buffer[T] {
+
+  def used: Int = q.used
+  def isFull: Boolean = q.isFull
+  def isEmpty: Boolean = q.isEmpty
+  def nonEmpty: Boolean = q.nonEmpty
+
+  def enqueue(elem: T): Unit = q.enqueue(elem)
+  def dequeue(): T = q.dequeue()
+
+  def peek(): T = q.peek()
+  def clear(): Unit = q.clear()
+  def dropHead(): Unit = q.dropHead()
+  def dropTail(): Unit = q.dropTail()
+
+  private final class FixedQueue extends Buffer[T] {
+    import Buffer._
+
+    private val queue = new Array[AnyRef](FixedQueueSize)
+    private var head = 0
+    private var tail = 0
+
+    override def capacity = BoundedBuffer.this.capacity
+    override def used = tail - head
+    override def isFull = used == capacity
+    override def isEmpty = tail == head
+    override def nonEmpty = tail != head
+
+    override def enqueue(elem: T): Unit =
+      if (tail - head == FixedQueueSize) {
+        val queue = new DynamicQueue(head)
+        while (nonEmpty) {
+          queue.enqueue(dequeue())
+        }
+        q = queue
+        queue.enqueue(elem)
+      } else {
+        queue(tail & FixedQueueMask) = elem.asInstanceOf[AnyRef]
+        tail += 1
+      }
+    override def dequeue(): T = {
+      val pos = head & FixedQueueMask
+      val ret = queue(pos).asInstanceOf[T]
+      queue(pos) = null
+      head += 1
+      ret
+    }
+
+    override def peek(): T =
+      if (tail == head) null.asInstanceOf[T]
+      else queue(head & FixedQueueMask).asInstanceOf[T]
+    override def clear(): Unit =
+      while (nonEmpty) {
+        dequeue()
+      }
+    override def dropHead(): Unit = dequeue()
+    override def dropTail(): Unit = {
+      tail -= 1
+      queue(tail & FixedQueueMask) = null
+    }
+  }
+
+  private final class DynamicQueue(startIdx: Int) extends ju.LinkedList[T] with Buffer[T] {
```

Comment:
I don't really like this implementation. This means that for large buffers we will suddenly put huge pressure on the heap and pay tons of allocation costs.

What I would like to see instead:
 - a top level array pointing to sub-arrays of size let's say 128, 128, 256, 512, ... up to some maximum capacity. On need of growth a new top-level array should be allocated and cirularity restored. On falling below a threshold, buffers should be dropped and circularity restored.

In the above scheme we can set the threshold to much lower and still have well behaved heap and allocation behavior, paying mostly for used buffer space (up to 2x), having better cache friendliness etc.